### PR TITLE
(PE-30410) Shadow fact/variable collisions in the PAL for plans

### PR DESCRIPTION
**Pull request by ekinanp**
This takes advantage of the recently merged Puppet 7 PR (https://github.com/puppetlabs/puppet/pull/8412)

Signed-off-by: Enis Inan <enis.inan@puppet.com>

### CODE REVIEW

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 1]
```diff
@@ -894,6 +894,7 @@
    (schema/required-key "trusted_facts") {(schema/required-key "values") {schema/Str schema/Any}}
    (schema/required-key "facts") {(schema/required-key "values") {schema/Str schema/Any}}
    (schema/required-key "variables") {(schema/required-key "values") (schema/either [{schema/Str schema/Any}] {schema/Str schema/Any})}
+   (schema/optional-key "target_variables") {(schema/required-key "values") [{schema/Str schema/Any}]}
```
Comment (ekinanp) [inline src/clj/puppetlabs/services/master/master_core.clj:897]: Made this optional for backwards compatibility. Also chose the `Array[Hash]` syntax since that seems to be the preferred way of specifying target var values.

Note that the Orch-side changes are basically to pass-in the `target_variables` to the request. I _think_ it consists of eliminating https://github.com/puppetlabs/orchestrator/blob/main/src/puppetlabs/orchestrator/bolt/plan_actions.clj#L335 and letting the compile endpoint handle the shadowing (and it will only be done on the `main` branch).

Comment (ekinanp) [reply src/clj/puppetlabs/services/master/master_core.clj:897]: I've tested that `target_variables` works as expected manually via `curl`.

File: src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb [hunk 2]
```diff
@@ -51,11 +51,23 @@ def self.compile_ast(code, compile_options, boltlib_path)
...
 
+          target_variables = if compile_options.key?('target_variables')
+                               compile_options['target_variables']['values'].each_with_object({}) do |param_hash, acc|
+                                 acc[param_hash.keys.first] = param_hash.values.first
+                               end
+                             else
+                               {}
+                             end
+
+          variables = {
```
Comment (ekinanp) [inline src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb:62]: Inspired by https://github.com/puppetlabs/bolt/pull/2277

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb [hunk 3]
```diff
@@ -70,5 +75,70 @@ def find_notify(catalog)
...
+        target_variables = {
+          "values" => [
+            # should be shadowed by target_var_fact fact
+            {"target_var_fact" => "target_value"},
+            # should be shadowed by plan_var_plain plan variable
+            {"plan_var_plain"  => "target_value"},
+          ]
+        }
+
+        code = <<CODE
```
Comment (ekinanp) [inline spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:114]: Based on https://github.com/puppetlabs/puppet/blob/main/lib/puppet/pal/pal_impl.rb#L516-L519, I think these cases should test everything.

File: src/clj/puppetlabs/services/master/master_core.clj (hunk 1 above)
Comment (ekinanp) [reply src/clj/puppetlabs/services/master/master_core.clj:897]: Where would I make the relevant doc updates for this?

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb [hunk 4]
```diff
@@ -70,5 +75,70 @@ def find_notify(catalog)
...
     end
+
+    context 'when compiling for a plan with fact and plan/target variable collisions' do
+      let(:boltlib_path) { [] }
+
+      it 'properly shadows the relevant variables' do
+        # TODO: This is a temporary stub of the Bolt requires to get this
+        # test to pass. We should figure out how to properly load Bolt code
+        # later for the tests.
+        allow(Puppet::Server::ASTCompiler).to receive(:require).and_call_original
```
Comment (ekinanp) [inline spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:86]: I know this sucks, but unless there's a _very_ good reason for us not to mock `require` in the tests, I'd rather not spend anymore time figuring out how to handle the Bolt requires in the AST compiler.

**Comment by donoghuc** [comment]
I really hope (and it looks like it is the case) that this landed in puppet6 based on https://github.com/puppetlabs/puppet/pull/8412#pullrequestreview-515880018 

We will need this in puppet6 to take advantage of this in both our stream.

File: src/clj/puppetlabs/services/master/master_core.clj (hunk 1 above)
Comment (donoghuc) [reply src/clj/puppetlabs/services/master/master_core.clj:897]: https://github.com/puppetlabs/osp-docs/blob/b0af18f7c554097a54e36f22d0403c592fe48272/server/puppet-api/v3/compile.md

**Comment by donoghuc** [comment]
Cool. so this did land in the 6.x stream it is just not in a released version yet. So this is fine.

File: src/clj/puppetlabs/services/master/master_core.clj (hunk 1 above)
Comment (donoghuc) [reply src/clj/puppetlabs/services/master/master_core.clj:897]: IMO the `plan_compile` specific options do not need to maintain backwards compatibility.

File: src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb (hunk 2 above)
Comment (donoghuc) [reply src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb:62]: This is fine. We could nitpick on the names but this is what bolt has done.

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 4 above)
Comment (donoghuc) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:86]: I briefly looked at this. When i did found that it was actually a pretty small subset of all the bolt dependency gems that we actually would need. I think so far we have relied heavily on testing this in orchestrator acceptance tests.

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb [hunk 5]
```diff
@@ -25,12 +25,17 @@ def request(code)
...
-def find_notify(catalog)
+def find_notify(catalog, title: nil)
   catalog['resources'].find do |item|
-    item['type'] == 'Notify'
+    next item['type'] == 'Notify' unless title
+    item['type'] == 'Notify' && item['title'] == title
   end
 end
 
+def collision_warning(msg_regex)
```
Comment (donoghuc) [inline spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:35]: I dont see where this is used?

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 4 above)
Comment (ekinanp) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:86]: Yeah, I'm going to add Orch acceptance tests for the shadowing if none currently exist. These tests at least are an easy way for us to detect a regression there if it happens so still useful to have even if the mocking currently sucks.

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 5 above)
Comment (ekinanp) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:35]: This was from a previous version of the unit tests when I was looking through the logs. Removed.

**Comment by mwaggett** [comment]
@ekinanp I think this should target 6.x, since the associated work is landing in the 2019.8.x stream.

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 4 above)
Comment (nicklewis) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:82]: How about we pull out the `require`s into a single `require_bolt` method we can stub normally?

Comment (donoghuc) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:82]: works for me. I'll make that update. good idea as usual @nicklewis :100:

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 3 above)
Comment (McdonaldSeanp) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:109]: Would there be any reason to cover the case that a var is shadowed more than once (i.e. set in all three of target var/plan var/fact)

Comment (McdonaldSeanp) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:109]: Would there be any reason to cover the case that a var is shadowed more than once (i.e. set in all three of target var/plan var/fact)

**Review by McdonaldSeanp — APPROVED** [review]
I do have one question, but it's a narrow test case so I still approve.

**Review by donoghuc — APPROVED** [review]

File: spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb (hunk 3 above)
Comment (donoghuc) [reply spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb:109]: No. we have that covered in orch and in bolt.
